diff -Naur openssl-3.3.1-prev/apps/cms.c openssl-3.3.1/apps/cms.c
--- openssl-3.3.1-prev/apps/cms.c	2024-08-10 21:22:08.633801154 +1000
+++ openssl-3.3.1/apps/cms.c	2024-08-10 21:43:13.210514693 +1000
@@ -84,7 +84,8 @@
     OPT_PROV_ENUM, OPT_CONFIG,
     OPT_V_ENUM,
     OPT_CIPHER,
-    OPT_ORIGINATOR
+    OPT_ORIGINATOR,
+    OPT_CRLFILE
 } OPTION_CHOICE;
 
 const OPTIONS cms_options[] = {
@@ -176,6 +177,7 @@
     {"md", OPT_MD, 's', "Digest algorithm to use"},
     {"signer", OPT_SIGNER, 's', "Signer certificate input file"},
     {"certfile", OPT_CERTFILE, '<', "Other certificates file"},
+    {"CRLfile", OPT_CRLFILE, '<', "Other CRLs file"},
     {"cades", OPT_CADES, '-',
      "Include signingCertificate attribute (CAdES-BES)"},
     {"nodetach", OPT_NODETACH, '-', "Use opaque signing"},
@@ -299,6 +301,7 @@
     long digestlen = 0;
     char *infile = NULL, *outfile = NULL, *rctfile = NULL;
     char *passinarg = NULL, *passin = NULL, *signerfile = NULL;
+    char *crlfile = NULL;
     char *originatorfile = NULL, *recipfile = NULL, *ciphername = NULL;
     char *to = NULL, *from = NULL, *subject = NULL, *prog;
     cms_key_param *key_first = NULL, *key_param = NULL;
@@ -474,6 +477,9 @@
         case OPT_CERTFILE:
             certfile = opt_arg();
             break;
+        case OPT_CRLFILE:
+            crlfile = opt_arg();
+            break;
         case OPT_CAFILE:
             CAfile = opt_arg();
             break;
@@ -1197,6 +1203,16 @@
                                        indata, out, flags))
             goto end;
     } else if (operation == SMIME_VERIFY) {
+        STACK_OF(X509_CRL) *crls = NULL;
+        if (crlfile) {
+            if (!load_crls(crlfile, &crls, NULL, "other CRLs")) {
+                exit(1);
+            }
+            for (int i = 0; i < sk_X509_CRL_num(crls); i++) {
+                X509_CRL *item = sk_X509_CRL_value(crls, i);
+                CMS_add1_crl(cms, item);
+            }
+        }
         if (CMS_verify(cms, other, store, indata, out, flags) > 0) {
             BIO_printf(bio_err, "%s Verification successful\n",
                        (flags & CMS_CADES) != 0 ? "CAdES" : "CMS");
